File: entrega_1/pso_simples/swarm.cpp
#include "swarm.h"

#include <cstdlib>
#include <math.h>
Swarm::Swarm(int partAmount, int dimAmount, int maxEpoch,
             TopologyFactory::TOPOLOGY top, AlgorithmFactory::ALGORITHM alg, VelocityFactory::TYPE velType)
{
    this->particleAmount = partAmount;
    this->dimensionAmount = dimAmount;
    this->maxEpoch = maxEpoch;

    AlgorithmFactory af(alg);
    this->algorithm = af.get();

    TopologyFactory tf(top);
    this->topology = tf.get();
    this->topology->init(partAmount);

    VelocityFactory vf(velType);
    this->velocity = vf.get();

    this->ran = new IagoRandom();

    initClerc();
}

Swarm::~Swarm()
{
    for(unsigned int i=0, size = this->particles->size(); i<size; ++i)
    {
        delete this->particles->at(i);
    }

    delete this->particles;
    delete this->ran;
    delete this->algorithm;
    delete this->topology;
    delete this->velocity;
}

std::vector<double> *Swarm::psoAlg()
{
    std::vector<double>* _outputFit = new std::vector<double>();
    //inicializando variáveis e partículas
    this->initialize();

    int size = this->particles->size();
    do
    {
        //vai parar se chegar no limite de epoch
        for(int i=0; i<size; ++i)
        {
            Particle *current = this->particles->at(i);
            double fit = this->algorithm->fitness(current, size);
            if(fit < current->getFitness())
            {
                current->setFitness(fit);
                current->setPBest(current->getData());//clona o vetor atual para pBest
            }
            this->topology->addFit(i, fit);
        }

        this->topology->updateBest(this->particles);
        _outputFit->push_back(this->topology->getBest());

        //para todas partículas
        for(int i=0; i<size; ++i)
        {
            Particle* current = this->particles->at(i);
            //atualiza o best da partícula
            this->topology->setBest(i, this->particles);
            for(int j=0; j<this->dimensionAmount; ++j)
            {
                this->calculateVelocityW(current, j);
                this->updatePosition(current, j);
            }
        }
        epoch++;
    }
    while(this->epoch < this->maxEpoch);
    return _outputFit;
}

void Swarm::initialize()
{
    this->epoch = 0;
    this->done = false;
    //this->bestFitness = RAND_MAX;

    this->algorithm->init(this->rangeMin, this->rangeMax, this->maxVelocity);

    this->particles = new std::vector<Particle*>();

    //criando as partículas
    for(int i=0; i<this->particleAmount; ++i)
    {
        Particle* p = new Particle(this->dimensionAmount);
        //criando valores para as dimensões
        for(int j=0; j<dimensionAmount; ++j)
        {
            //inicializando dentro dos limites
            int randomVal = ran->randRange(this->rangeMin, this->rangeMax);
            p->setData(j, randomVal);
            this->randomSpeed(p);
        }
        p->setPBest(p->getData());
        this->particles->push_back(p);
    }
}

void Swarm::calculateVelocity(Particle *part, int pos)
{
    double c1 = 2.05;
    double c2 = 2.05;

    double rand1 = ran->randNormal();
    double rand2 = ran->randNormal();

    double velocity = (part->getVelocity(pos) +
                           (c1 * rand1 * (part->getPBest(pos) - part->getData(pos))) +
                           (c2 * rand2 * (part->getGBest()->at(pos) - part->getData(pos))));

    //clamping
    if(velocity > this->maxVelocity)
        velocity = this->maxVelocity;
    if(velocity < -this->maxVelocity)
        velocity = -this->maxVelocity;
    part->setVelocity(pos, velocity);
}

void Swarm::calculateVelocityW(Particle *part, int pos)
{
    double w = 0.8;
    double c1 = 2.05;
    double c2 = 2.05;

    double rand1 = ran->randNormal();
    double rand2 = ran->randNormal();

    double velocity = w * (part->getVelocity(pos) +
                           (c1 * rand1 * (part->getPBest(pos) - part->getData(pos))) +
                           (c2 * rand2 * (part->getGBest()->at(pos) - part->getData(pos))));

    //clamping
    if(velocity > this->maxVelocity)
        velocity = this->maxVelocity;
    if(velocity < -this->maxVelocity)
        velocity = -this->maxVelocity;
    part->setVelocity(pos, velocity);
}

void Swarm::calculateVelocityDec(Particle *part, int pos)
{
    //w inicial
    double w = 0.9;
    //w final
    double endW = 0.4;

    //tamanho da reduçao por iteraçao
    double wStep = endW/this->maxEpoch;

    //w atual apos reduçao por iteraçoes
    double currentW = w - (epoch*wStep);

    double c1 = 2.05;
    double c2 = 2.05;

    double rand1 = ran->randNormal();
    double rand2 = ran->randNormal();

    double velocity = currentW*part->getVelocity(pos) +(
                                 (c1 * rand1 * (part->getPBest(pos) - part->getData(pos))) +
                                 (c2 * rand2 * (part->getGBest()->at(pos) - part->getData(pos)))
                );

    //clamping
    if(velocity > maxVelocity)
        velocity = maxVelocity;

    if(velocity < -maxVelocity)
        velocity = -maxVelocity;

    part->setVelocity(pos, velocity);
}


void Swarm::initClerc()
{
    this->_c1 = ran->randRange(1,4);
    _c2 = 4.1 - _c1;
    this->_phi = _c1 + _c2;

    double root = sqrt((_phi*_phi)-(4*_phi));
    double bot = 2 - _phi - root;
    if(bot < 0) bot = -bot;
    _X = 2/bot;
}

void Swarm::calculateVelocityC(Particle *part, int pos)
{
    double rand1 = ran->randNormal();
    double rand2 = ran->randNormal();

    double velocity = _X*(part->getVelocity(pos) +
                                 (_c1 * rand1 * (part->getPBest(pos) - part->getData(pos))) +
                                 (_c2 * rand2 * (part->getGBest()->at(pos) - part->getData(pos))));

    part->setVelocity(pos, velocity);
}

void Swarm::updatePosition(Particle *part, int pos)
{
    double vel = part->getVelocity(pos);//velocidade já atualizada
    double oldPos = part->getData(pos);

    part->setData(pos, vel + oldPos);
}

void Swarm::randomSpeed(Particle *part)
{
    int size = this->dimensionAmount;
    for(int i=0; i<size; ++i)
    {
        double randSpeed = ran->randRange(-4,4);
        if(randSpeed==0)randSpeed++;
        part->setVelocity(i, randSpeed);
    }
}

double Swarm::sumVector(std::vector<double> *v)
{
    double out = 0;
    for(double d : *v)
        out += d;
    return out;
}

File: entrega_5/Java/src/main_src/Swarm.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package main_src;

import algorithm.AlgorithmEnum;
import algorithm.AlgorithmFactory;
import algorithm.IAlgorithm;
import java.time.Clock;
import java.util.ArrayList;
import java.util.concurrent.ThreadLocalRandom;
import topology.ITopology;
import topology.TopologyEnum;
import topology.TopologyFactory;
import velocity.IVelocity;
import velocity.VelocityEnum;
import velocity.VelocityFactory;

/**
 *
 * @author Geraldo e Iago
 */
public class Swarm
{
    private int particleAmount;
    private int dimensionAmount;
    private int maxEpoch;
    
    private ArrayList<Particle> particles;
    
    private int epoch;
    
    private ITopology topology;
    private IAlgorithm algorithm;
    private IVelocity velocity;
    
    public Swarm(int partAmount, int dimAmount, int maxEpoch, TopologyEnum top,
            AlgorithmEnum alg, VelocityEnum vel)
    {
        this.particleAmount = partAmount;
        this.dimensionAmount = dimAmount;
        this.maxEpoch = maxEpoch;
        
        TopologyFactory topFac = new TopologyFactory(top);
        this.topology = topFac.get();
        this.topology.init(partAmount);
        
        AlgorithmFactory algFac = new AlgorithmFactory(alg);
        this.algorithm = algFac.get();
        
        VelocityFactory velFac = new VelocityFactory(vel);
        this.velocity = velFac.get();
    }
    
    public ArrayList psoAlg()
    {
        ArrayList outputFit = new ArrayList();
        ArrayList positions = new ArrayList();
        //inicializando
        this.initialize();
        
        int size = this.particles.size();
        do
        {
            double lastFit = Double.MAX_VALUE;
            if(outputFit.size()>0)
                lastFit = (double) outputFit.get(outputFit.size()-1);
            //Temos que calcular a soma dos fitness individuais
            //pra ver se a melhora foi geral
            double fitnessSum = 0;
            //calcula o fitness individual
            for(int i=0; i<size; i++)
            {
                Particle current = this.particles.get(i);
                double fit = this.algorithm.fitness(current, particles);
                if(fit< current.getFitness())
                {
                    current.setFitness(fit);
                    current.setPBest();
                }
                //this.topology.setFitAt(i, fit);
                //somando ao fitness dessa iteração
                fitnessSum += fit;
            }
            
            
            /*
            Se o fitnesse melhorou, então:
            atualize o melhor
            armazene
            calcule velocidade
            movimente
            */
            if(fitnessSum < this.topology.getBest())
            {
                lastFit = fitnessSum;
                this.topology.setGBest(fitnessSum, particles);
                for(int i=0; i<size; i++)
                {
                    Particle current = this.particles.get(i);

                    for(int j=0; j<this.dimensionAmount; j++)
                    {
                        //this.calculateVelocityW(current, j);
                        this.velocity.calculate(current, j, this.algorithm.getVMax(), epoch, maxEpoch);
                        this.updatePosition(current, j);
                    }
                }
                positions.add(particlesToString());
            }
            outputFit.add(lastFit);
            
            //movimentação aleatória
            for(int i=0; i<size; i++)
            {
                Particle current = this.particles.get(i);
                randomSpeed(current);
                for(int j=0; j<this.dimensionAmount; j++)
                {
                    randomSpeed(current);
                    this.updatePosition(current, j);
                }
            }
            
            
            epoch++;
        }while(this.epoch < this.maxEpoch);//para no limite || trocar para condição do problema, i.e. N ciclos parado
        
        /* imprimindo a variação de fit */
        String out = "";
        for(int i=0, s = outputFit.size(); i<s; i++)
        {
            out += outputFit.get(i)+"\n";
        }
        out = out.substring(0, out.length()-1);
        ProjectIO io = new ProjectIO();
        io.writeFile("fit.csv", out);
        
        return positions;
    }
    
    private String particlesToString()
    {
        String out = "";
        for(int i=0, size = this.particles.size(); i<size; i++)
        {
            out += this.particles.get(i).toString()+";";
        }
        out = out.substring(0, out.length()-1);
        return out;
    }
    
    private double random(double max, double min)
    {
        return ThreadLocalRandom.current().nextDouble((int)min, (int)max);
    }
    
    private void randomSpeed(Particle part)
    {
        int size = this.dimensionAmount;
        for(int i=0; i<size; i++)
        {
            double _max = this.algorithm.getVMax();
            double randomSpeed = random(_max, -_max);
            if(randomSpeed==0) randomSpeed++;
            part.setVelocity(i, randomSpeed);
        }
    }
    
    /**
     *  Inicializa as partículas em posições aleatórias e com velocidades aleatórias.
     */
    private void initialize()
    {
        this.epoch = 0;
        this.algorithm.init();
        
        ProjectIO io = new ProjectIO();
        ArrayList<Disease> diseases = io.readFile("entrada.csv");
        
        int id = 0;
        this.particles = new ArrayList<>();
        for(int i=0, size = diseases.size(); i<size; i++)
        {
            Particle p = new Particle(diseases.get(i), this.dimensionAmount, id++);
            p = randomize(p);
            this.particles.add(p);
        }
        
        //criando as partículas
        for(int i=0, size=particleAmount-diseases.size(); i<size; i++)
        {
            int rand = (int) (Math.random()*diseases.size());
            Particle p = new Particle(diseases.get(rand), this.dimensionAmount, id++);
            p = randomize(p);
            this.particles.add(p);
        }
    }
    
    private Particle randomize(Particle p)
    {
        //posicionando aleatoriamente
        for(int j=0; j<dimensionAmount; j++)
        {
            int randomVal = (int)random(this.algorithm.getMax(), this.algorithm.getMin());
            p.setData(j, randomVal);
            p.setPBest(j, randomVal);//o melhor é o inicial
        }
        this.randomSpeed(p);
        
        return p;
    }
    /*
    private void calculateVelocityW(Particle part, int pos)
    {
        double w= 0.8;
        double c1 = 2.05;
        double c2 = 2.05;
        
        double rand1 = random(1, 0);
        double rand2 = random(1, 0);
        
        double v1 = c1 * rand1 * (part.getPBest(pos) - part.getData(pos));
        double v2 = c2 * rand2 * (part.getGBest(pos) - part.getData(pos));
        double _velocity = w * part.getVelocity(pos) + v1 + v2;
        
        //clamping
        double max = this.algorithm.getVMax();
        if(_velocity> max)
            _velocity = max;
        if(_velocity< -max)
            _velocity = -max;
        
        part.setVelocity(pos, _velocity);
    }
    */
    private void updatePosition(Particle part, int pos)
    {
        double vel = part.getVelocity(pos);//valor já atualizado
        double oldPos = part.getData(pos);
        
        double newPos = vel+oldPos;
        if(newPos < this.algorithm.getMin())
            newPos = this.algorithm.getMin();
        if(newPos > this.algorithm.getMax())
            newPos = this.algorithm.getMax();
        
        part.setData(pos, newPos);
    }
}